packages/vscode-extension/src/telemetry/telemetry.ts

```diff
@@ -0,0 +1,339 @@
+import * as vscode from "vscode";
+
+import {
+  ConfigFolderName,
+  FxError,
+  Stage,
+  TelemetryReporter,
+  UserError,
+} from "@microsoft/teamsfx-api";
+import { Correlator } from "@microsoft/teamsfx-core";
+import Reporter from "@vscode/extension-telemetry";
+
+import * as extensionPackage from "../../package.json";
+import * as globalVariables from "../globalVariables";
+import {
+  FeatureFlags,
+  getAllFeatureFlags,
+  getProjectId,
+  isFeatureFlagEnabled,
+} from "../utils/commonUtils";
+import { TelemetryCache } from "./cache";
+import {
+  TelemetryComponentType,
+  TelemetryErrorType,
+  TelemetryEvent,
+  TelemetryEventCache,
+  TelemetryProperty,
+  TelemetrySuccess,
+} from "./extTelemetryEvents";
+
+const TelemetryTestLoggerFile = "telemetryTest.log";
+
+/* eslint-disable prefer-const */
+let isFromSample: boolean | undefined = undefined;
+let settingsVersion: string | undefined = undefined;
+let isM365: boolean | undefined = undefined;
+
+export let reporter: VSCodeTelemetryReporter;
+// export for UT
+export let lastCorrelationId: string | undefined = undefined;
+
+/**
+ *  VSCode telemetry reporter used by fx-core and extension.
+ *  Usage:
+ *    let reporter = new VSCodeTelemetryReporter(key, extensionVersion, extensionId)
+ *  Illustrate:
+ *    key = <'the application insights key'>, 'aiKey' in package.json
+ *    extensionVersion = '<extension version>', extension version will be reported as a property with each event
+ *    extensionId = '<your extension unique name>', all events will be prefixed with this event name. eg: 'extensionId/eventname'
+ */
+export class VSCodeTelemetryReporter extends vscode.Disposable implements TelemetryReporter {
+  private readonly reporter: Reporter;
+  private readonly logger: Logger | undefined;
+  private readonly testFeatureFlag: boolean;
+  /**
+   * events are cached in memory first and sent later to reduce date loss.
+   * strategy: send at least once.
+   * assumption: if an event is sent to server 10 seconds ago, it is treated as successfully sent telemetry event.
+   */
+  private cache: TelemetryCache;
+
+  private sharedProperties: { [key: string]: string } = {};
+
+  constructor(key: string, extensionVersion: string, extensionId: string) {
+    super(async () => await this.reporter.dispose());
+    this.reporter = new Reporter(extensionId, extensionVersion, key, true);
+    this.testFeatureFlag = isFeatureFlagEnabled(FeatureFlags.TelemetryTest);
+    if (this.testFeatureFlag) {
+      const logFile = path.join(os.homedir(), `.${ConfigFolderName}`, TelemetryTestLoggerFile);
+      configure({
+        appenders: { everything: { type: "file", filename: logFile } },
+        categories: { default: { appenders: ["everything"], level: "debug" } },
+      });
+      this.logger = getLogger("TelemTest");
+    }
+    this.cache = new TelemetryCache(this.reporter);
+  }
+
+  public addSharedProperty(name: string, value: string): void {
```

Comment:
Will this refactor affect AB testing scorecard since exp service is still using the clientTimestamp for analysis?

Reply:
As discussed, it should have little impact when grouping event time. Another impact is duplicated events which will affect both control and treatment group.